src/js/builtins/Module.ts

```diff
@@ -34,53 +43,83 @@ export function overridableRequire(this: CommonJSModuleRecord, id: string) {
     $evaluateCommonJSModule(existing);
     return existing.exports;
   }
-
   if (id.endsWith(".node")) {
     return $internalRequire(id);
   }
-
   // To handle import/export cycles, we need to create a module object and put
   // it into the map before we import it.
   const mod = $createCommonJSModule(id, {}, false, this);
   $requireMap.$set(id, mod);
 
+  let { Module } = cache;
+  if (Module === undefined) {
+    // @ts-ignore - TS doesn't know each function is isolated.
+    Module = require("node:module");
+    cache.Module = Module;
+  }
+  const { _extensions } = Module;
+  if (_extensions && _extensions.$structureChanged) {
+    let { basename } = cache;
+    if (basename === undefined) {
+      // @ts-ignore - TS doesn't know each function is isolated.
+      ({ basename } = require("node:path"));
+    }
+    const name = basename(id);
+    let index = 0;
+    let startIndex = 0;
+    // Find longest registered extension.
+    while ((index = $stringIndexOfInternal.$call(name, '.', startIndex)) !== -1) {
+      const ext = index 
+        ? $stringSubstring.$call(name, index) 
+        // Skip dotfiles like .gitignore
+        : '';
+      // Defer to custom extension handler if present.
+      if (ext && typeof _extensions[ext] === 'function') { 
+        _extensions[ext](mod, id);
+        return mod.exports;
+      }
+      startIndex = index + 1;
+    }
+    // Fallback to ".js".
+    if (typeof _extensions[".js"] === 'function') {
+      _extensions[".js"](mod, id);
+      return mod.exports;
+    }
+  }
   // This is where we load the module. We will see if Module._load and
   // Module._compile are actually important for compatibility.
   //
   // Note: we do not need to wrap this in a try/catch, if it throws the C++ code will
   // clear the module from the map.
-  //
-  var out = this.$require(
-    id,
+  if (this.$require(id,
     mod,
-    // did they pass a { type } object?
+    // Did they pass a { type } object?
+    // Use `@argumentCount()` to avoid cloned arguments allocation.
     $argumentCount(),
-    // the object containing a "type" attribute, if they passed one
+    // The object containing a "type" attribute, if they passed one
     // maybe this will be "paths" in the future too.
-    arguments[1],
-  );
-
-  // -1 means we need to lookup the module from the ESM registry.
-  if (out === -1) {
+    type,
```

Comment:
Uses a param instead of materializing the `arguments` object. To better take advantage of the `@argumentCount` use.